f3.go

```diff
@@ -145,412 +78,197 @@ func New(ctx context.Context, manifest manifest.ManifestProvider, ds datastore.D
-	case err := <-runnerErrCh:
-		return err
-	case err := <-manifestErrCh:
+func (m *F3) Start(startCtx context.Context) (_err error) {
+	err := m.manifestProvider.Start(startCtx)
+	if err != nil {
 		return err
 	}
-}
-
-func (m *F3) Stop() {
-	m.mLk.Lock()
-	defer m.mLk.Unlock()
 
-	m.pauseRunner()
-
-	if m.cancelCtx != nil {
-		m.cancelCtx()
+	// Try to get an initial manifest immediately if possible so uses can query it immediately.
+	var pendingManifest *manifest.Manifest
+	select {
+	case pendingManifest = <-m.manifestProvider.ManifestUpdates():
+		m.mu.Lock()
+		m.manifest = pendingManifest
+		m.mu.Unlock()
+	default:
 	}
-}
 
-func (m *F3) pauseRunner() {
-	if m.runner != nil {
-		m.runner.Stop()
+	initialDelay, err := m.computeBootstrapDelay(pendingManifest)
+	if err != nil {
+		return err
 	}
-	m.runner = nil
-}
 
-func (m *F3) handleIncomingMessages(ctx context.Context, queue chan gpbft.ValidatedMessage) {
-	ctx, m.client.incomingCancel = context.WithCancel(ctx)
-loop:
-	for {
-		var msg *pubsub.Message
-		msg, err := m.msgSub.Next(ctx)
-		if err != nil {
-			if ctx.Err() != nil {
-				break
-			}
-			m.log.Errorf("msgPubsub subscription.Next() returned an error: %+v", err)
-			break
-		}
-		gmsg, ok := msg.ValidatorData.(gpbft.ValidatedMessage)
-		if !ok {
-			m.log.Errorf("invalid msgValidatorData: %+v", msg.ValidatorData)
-			continue
-		}
-		select {
-		case queue <- gmsg:
-		case <-ctx.Done():
-			break loop
+	// Try to start immediately if we have a manifest available and don't have to wait to
+	// bootstrap. That way, we'll be fully started when we return from Start.
+	if initialDelay == 0 {
+		if err := m.reconfigure(startCtx, pendingManifest); err != nil {
+			return xerrors.Errorf("failed to start GPBFT: %w", err)
 		}
+		pendingManifest = nil
 	}
-}
 
-// Callback to be triggered when there is a dynamic manifest change
-// If the manifest triggers a rebootstrap it starts a new runner with the new configuration.
-// If there is no rebootstrap it only starts a new pubsub topic with a new network name that
-// depends on the manifest version so there is no overlap between different configuration instances
-func ManifestChangeCallback(m *F3) manifest.OnManifestChange {
-	// We can only accommodate 1 manifest update at a time, thus
-	// the buffer size.
-	manifestUpdate := make(chan uint64, 1)
-	m.client.manifestUpdate = manifestUpdate
-	return func(ctx context.Context, prevManifest manifest.Manifest, errCh chan error) {
-		// Tear down pubsub.
-		if err := m.teardownPubsub(prevManifest); err != nil {
-			// for now we just log the error and continue.
-			// This is not critical, but alternative approaches welcome.
-			m.log.Errorf("error stopping gpbft runner: %+v", err)
+	m.errgrp.Go(func() error {
+		manifestChangeTimer := time.NewTimer(initialDelay)
+		if pendingManifest == nil && !manifestChangeTimer.Stop() {
+			<-manifestChangeTimer.C
 		}
-		// empty message queue from outstanding messages
-		m.emptyMessageQueue()
```

Comment:
I re-create the message queue each time. This was also a bit racy as we could ahve processed one more message after closing pubsub.